src/ctap/command.rs

```diff
@@ -124,30 +124,32 @@ impl TryFrom<cbor::Value> for AuthenticatorMakeCredentialParameters {
     type Error = Ctap2StatusCode;
 
     fn try_from(cbor_value: cbor::Value) -> Result<Self, Ctap2StatusCode> {
-        let param_map = read_map(&cbor_value)?;
+        let mut param_map = extract_map(cbor_value)?;
 
-        let client_data_hash = read_byte_string(ok_or_missing(param_map.get(&cbor_unsigned!(1)))?)?;
+        let client_data_hash =
+            extract_byte_string(ok_or_missing(param_map.remove(&cbor_unsigned!(1)))?)?;
```

Comment:
Rather than removing elements one by one, which doesn't have ideal performance and by itself could copy values around inside the map, would it be possible to loop over `map.into_iter()` and rely on the fact that keys are sorted in a `BTreeMap`?

I'm thinking of something like the following (very rough draft), but maybe a macro would make it easier to use (a "mirror" of `cbor_map!` that creates CBOR values).

```rust
let mut map = extract_map(cbor_value)?;
let mut it = map.into_iter();

let client_data_hash = loop {
    let (k, v) = it.next()?;
    match k.cmp(cbor_unsigned!(1)) {
        Ordering::Less => /* ignore unknown key */ continue,
        Ordering::Equal => extract_byte_string(v),
        Ordering::Greater => /* key "1" is not in the map */ return Err(...),
    }
};

// Same with cbor_unsigned!(2).
// etc.
```

Reply:
Do you think that the proposed implementation is slower than the former one with `get`?

Reply:
I think that readability is more important than performance. In particular the performance benefit is very small here. The maps are small (around 10 elements on average). Re-balancing is thus cheap (it's just rewriting 5 or so pointers) and on par with the cost of moving the value (that we have to do anyway). So we won't gain anything significant.

Reply:
If performance is not so important, what is the rationale for moving rather than copying? The current implementation with `get` is readable enough and the few copies don't have that much overhead.

And on the readability front, my idea would be to embed this logic in a macro, so that it's readable. For example:

```rust
// This internally has `let client_data_hash = ...` statements creating the `client_data_hash`,
// `rp`, ... variables.
parse_cbor_map!{param_map,
    1 => client_data_hash,
    2 => rp,
    3 => user,
    ...
};

let client_data_hash = extract_byte_string(client_data_hash)?;
let rp = PublicKeyCredentialRpEntity::try_from(rp)?;
...
```

Then we can easily swap the internal implementation of this `parse_cbor_map!` macro, to experiment with `get`, `remove` or `into_iter` implementations.

Reply:
Yes, such a macro would definitely help and we should probably add it (this is the kind of stuff you get for free with protocol buffers, serde, or other libraries based on schema).

The rationale I see for this change is that parsing is something that is essentially linear and reflecting that in the types is natural. This is surprising to see a copy-based parsing libraries. So the change is not about performance but about API.

Reply:
What this PR is supposed to do is clean up from former PRs where Julian used moves in data_formats.rs, and have a consistent way of parsing the data structures. Can we postpone any potential performance improvements to a later PR? Your latest thumbs up lets me assume that we like the remove better than the get for now?

Reply:
My take is mostly that there is an extra cost to refactoring all the use cases from get to remove, and then from remove to the macro, as opposed to refactoring directly from get to the macro, and then changing how the macro is implemented.

That said, now that the get -> remove refactoring is implemented, the cost to moving to the macro will be essentially the same, whether this pull request is merged or not.

So I'm pretty neutral on this pull request. It's not a perfect performance improvement due to the remove being non-trivial, but it at least clarifies that parsing is linear (consuming types rather than taking references).

Reply:
I'll go ahead with it then, so we at least have only one of the two: `get`, `remove`